### ISSUE TITLE
Building Dict is too slow

### ISSUE THREAD

**Issue by shashankMadan-designEsthetics** [facebookresearch/ParlAI#2848]
**Bug description**
```
from parlai.scripts.train_model import TrainModel

TrainModel.main(
    # similar to before
    task='amazon_qa',
    model='projects.wizard_of_wikipedia.generator.agents:EndToEndAgent',
    model_file='/tmp/end2end_generator/model',

    # initialize with a pretrained model
    init_model='zoo:wizard_of_wikipedia/end2end_generator/model',

    # arguments we get from the pretrained model.
    # Unfortunately, these must be looked up separately for each model.
    # eps
    dict_lower=True,
    dict_tokenizer='bpe',
    n_layers=5,
    n_heads=2,
    dropout=0.20,
    ffn_size=512,
    embedding_size=256,
    log_every_n_secs=10,
    validation_patience=12,
    validation_metric='ppl',
    validation_metric_mode='min',
    validation_every_n_epochs=0.5,
    n_positions=128,
    truncate=128,
    max_knowledge=32,
    knowledge_alpha=0.95,
    knowledge_truncate=32,
    learningrate=5e-4,
    warmup_updates=5000,
    clip=0.1,
    lr_scheduler='invsqrt',
    embedding_type='fasttext',
    beam_size=1,
    skip_generation=False,
    batchsize=64,
)

```
I am trying to train amazon_qa task on wizard of Wikipedia model, just to experiment it out, I am not sure if it will work but when I run this script it says creating a task and goes to next stage building dictionary it just becomes to slow of a process.

<img width="1021" alt="Screenshot 2020-07-18 at 10 42 32 PM" src="https://user-images.githubusercontent.com/45225143/87858114-2f34b380-c949-11ea-9928-3bfc77fa91c8.png">

Like has been around 2 hrs and it hasn't crossed 0% yet.
Can anyone please point me out the error, thanks.

**Comment by stephenroller** [comment]
Hmm, it seems like it crashed. My own building of a BPE dictionary runs at about 8000ex/s. Which would take 10 minutes to build, but you can cache it.

Can you run `python -m parlai.scripts.display_data -t amazon_qa` ?

**Comment by stephenroller** [comment]
Actually I found it. it's an issue with this particular dataset, `amazon_qa`. Will post a patch soon.

**Comment by stephenroller** [comment]
#2851 has a fix. My coworkers probably won't review it until Monday.

**Comment by shashankMadan-designEsthetics** [comment]
Great! hope it gets resolved...

**Comment by stephenroller** [comment]
You can apply the patch yourself.

Looks like that particular dataset has a few other oddities about it. It doesn't have a validation or test set, for example.

**Comment by shashankMadan-designEsthetics** [comment]
I did apply the patch myself... it works! thanks a lot @stephenroller 
And yes my data set doesn't have any splits I am new here so was just trying to get it to work. I'll look into the validation or test set. Do you have any references to give?
